### PR TITLE
HPCC-32724 Allow storage planes to be defined in bare-metal systems 

### PR DESCRIPTION
**Pull request by jackdelv**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: dali/base/dafdesc.cpp [hunk 1]
```diff
@@ -3699,12 +3699,20 @@ static void doInitializeStorageGroups(bool createPlanesFromGroups)
             if (cur.groupType == grp_thor)
                 cur.createStoragePlane(storage, 1);
         }
+    }
 
-        //Uncomment the following to trace the values that been generated
-        //printYAML(storage);
+    // Add groups defined in environment.xml
+    Owned<IPropertyTree> env = getHPCCEnvironment();
+    Owned<IPropertyTreeIterator> storagePlanes = env->getElements("Software/Globals/StoragePlaneList/StoragePlane[@name='*']");
```
Comment (ghalliday) [inline dali/base/dafdesc.cpp:3706]: minor:I would use "Storage" (or "StoragePlanes" rather than "StoragePlaneList" 

Also, no need to have a qualifier on the name:

```
Owned<IPropertyTreeIterator> storagePlanes = env->getElements("Software/Globals/StoragePlanes/StoragePlane");
```
I am not sure about the benefits of accessing via the environment, or by copying the data to the component configuration file and accessing from there.
@jakesmith opinions?

File: testing/regress/environment.xml.in [hunk 2]
```diff
@@ -20,6 +20,20 @@
         storageWrites="0.0038"/>
  </Hardware>
  <Software>
+  <Globals noCommon="true"
+           maxConnections="-1">
+           <StoragePlane name="storagePlane"
```
Comment (ghalliday) [inline testing/regress/environment.xml.in:25]: Doesn't contain StoragePlaneList

Also, I suspect these should not be checked in as part of the PR.

File: common/remote/hooks/azure/azurefile.cpp [hunk 3]
```diff
@@ -459,12 +459,15 @@ AzureFile::AzureFile(const char *_azureFileName) : fullName(_azureFileName)
 
         StringBuffer planeName(slash-filename, filename);
         Owned<IPropertyTree> plane = getStoragePlane(planeName);
+        if (!plane)
```
Comment (ghalliday) [inline common/remote/hooks/azure/azurefile.cpp:462]: Should be in a separate PR.

File: common/remote/hooks/azure/azurefile.cpp [hunk 4]
```diff
@@ -459,12 +459,15 @@ AzureFile::AzureFile(const char *_azureFileName) : fullName(_azureFileName)
...
+            throw makeStringExceptionV(99, "Unknown storage plane %s", planeName.str());
+
         const char * api = plane->queryProp("storageapi/@type");
         if (!api)
             throw makeStringExceptionV(99, "No storage api defined for plane %s", planeName.str());
 
         constexpr size_t lenPrefix = strlen(azureBlobPrefix);
-        if ((strncmp(api, azureBlobPrefix, lenPrefix-1) != 0) || api[lenPrefix-1] != ':')
+        if ((strncmp(api, azureBlobPrefix, lenPrefix-1) != 0))
```
Comment (ghalliday) [inline common/remote/hooks/azure/azurefile.cpp:470]: I think the extra test should have been
```
api[lenPrefix-1] == 0
```
so that it doesn't match azureblobv2

possibly better would be to define a string without the trailing colon and use strieq.

**Review by ghalliday — COMMENTED** [review]
Changes look reasonable.  Need to separate out the different types of changes though.

**Comment by jackdelv** [comment]
@ghalliday I have removed changes unrelated to adding support for storage planes in bare-metal. The working code and the commits based on your comments can be found on [branch HPCC-32723](https://github.com/hpcc-systems/HPCC-Platform/compare/master...jackdelv:HPCC-Platform:HPCC-32723?expand=1) in my forked repo.

File: common/remote/hooks/azure/azurefile.cpp (hunk 4 above)
Comment (jackdelv) [reply common/remote/hooks/azure/azurefile.cpp]: Changed to use strieq.

File: common/remote/hooks/azure/azurefile.cpp (hunk 3 above)
Comment (jackdelv) [reply common/remote/hooks/azure/azurefile.cpp]: Moved to HPCC-32723 on my local repo.

File: testing/regress/environment.xml.in (hunk 2 above)
Comment (jackdelv) [reply testing/regress/environment.xml.in]: Removed initfiles changes from PR.

**Review by ghalliday — APPROVED** [review]

**Comment by ghalliday** [comment]
@jackdelv please convert to normal PR
@jakesmith can you foresee any problems with this?

File: dali/base/dafdesc.cpp (hunk 1 above)
Comment (jakesmith) [reply dali/base/dafdesc.cpp]: The standard approach is to generate into the component config file (in this case daliconf.xml).
We do this already for other components (dfuserver, esp.. ), see e.g. initfiles/componentfiles/configxml/dfuserver.xsl. At the moment it does this for RemoteStorage only in those places.

<planes> should be generated in the same way into those configs too, and Dali (which doesn't need RemoteStorage) should generate <planes> only.

One caveat to that is that the above code only creates planes for existing groups in Dali (that correspond to engine groups in BM) if "planes" doesn't exist, but it will do after generating it into the config..

I suggest changing the createStoragePlane calls, so that in this context it adds a flag to the planes being added, to indicate they are based on groups (e.g. @fromGroup=true), and then change the start of this process to:
```
    if (createPlanesFromGroups)
    {
        // delete all [@fromGroup=true] planes 
        
        // existing code that constructs groups and adds planes for groups


...
     }
```

NB: this routine can be recalled if/when environment.xml changes.

**Review by jakesmith — COMMENTED** [review]
@jackdelv - please see comments

File: dali/base/dafdesc.cpp (hunk 1 above)
Comment (jackdelv) [reply dali/base/dafdesc.cpp]: Copied storage/planes definitions into component config files which allowed a group to be created for my storage plane.

I changed createStoragePlane to add the flag fromGroup to the planes created from it. Then at the beginning of the process changed it to remove all storage planes previously created by createStoragePlane by removing any with fromGroup=true. I also had to add a check at the beginning of createStoragePlanes for a storage plane that is already defined (this would be coming from the component config file) otherwise it would create a duplicate planes.

File: dali/base/dafdesc.cpp [hunk 5]
```diff
@@ -3652,8 +3659,12 @@ static void doInitializeStorageGroups(bool createPlanesFromGroups)
         storage.set(globalConfig->addPropTree("storage"));
 
 #ifndef _CONTAINERIZED
-    if (createPlanesFromGroups && !storage->hasProp("planes"))
+    if (createPlanesFromGroups)
     {
+        // Remove old planes created from groups
+        while (storage->removeProp("planes[@fromGroup=\"1\"]"));
```
Comment (jakesmith) [inline dali/base/dafdesc.cpp:3665]: trivial: for consistency match with true vs 1, and can use single quotes for readability:
while (storage->removeProp("planes[@fromGroup='true']"));

**Review by jakesmith — APPROVED** [review]
@jackdelv - looks good. Please address 1 trivial comment and squash.

**Comment by jackdelv** [comment]
@jakesmith Squashed. I'm looking into why some of the smoketests are failing.

**Comment by ghalliday** [comment]
@jackdelv tests are still failing.  Did Jakes suggestion of changing from ='1' to ='true' break it?

**Review by ghalliday — CHANGES_REQUESTED** [review]
tests currently failing

**Comment by jackdelv** [comment]
@ghalliday  It seems that may have been the issue. I will wait for the tests to run to be sure since I'm still confused why it was happening. When I run this code for connecting to my azure blob storage it seemed to be working on roxie. It was hitting an unimplemented error rather than the missing host group exception.

> UNIMPLEMENTED feature [AzureFile::copyTo] in function copyTo() at azurefile.cpp(246)

```
in := DATASET('~plane::jackdelvstorageplanetest::taxi-blob::taxi_tripdata.csv', layout, CSV(HEADING(1)));
out := CHOOSEN(in,1000);
OUTPUT(out);
OUTPUT(out,,'~plane::jackdelvstorageplanetest::taxi-blob::taxi_tripdata_copy.csv',CSV, PLANE('jackdelvstorageplanetest'), OVERWRITE);
in2 := DATASET('~plane::jackdelvstorageplanetest::taxi-blob::taxi_tripdata_copy.csv', layout, CSV);
OUTPUT(in2,,'~mytaxi::taxi_tripdata_copy.csv',CSV, PLANE('jackdelvstorageplanetest'), OVERWRITE);
```

I tested with testing/regress/payload.ecl and was getting the same host group error with 'true'. Changing to '1' seemed to fix the error.

> EXCEPTION: No entry found for hostGroup: 'mythor' (in Index Read 2)

Although, when the test suite reported the error, it was looking for myroxie, but I could only ever reproduce with the mythor host group.

File: dali/base/dafdesc.cpp [hunk 6]
```diff
@@ -3652,8 +3659,12 @@ static void doInitializeStorageGroups(bool createPlanesFromGroups)
         storage.set(globalConfig->addPropTree("storage"));
 
 #ifndef _CONTAINERIZED
-    if (createPlanesFromGroups && !storage->hasProp("planes"))
+    if (createPlanesFromGroups)
     {
+        // Remove old planes created from groups
+        while (storage->removeProp("planes[@fromGroup='1']"));
```
Comment (ghalliday) [inline dali/base/dafdesc.cpp:3665]: minor style: For loops with empty bodies it is clearer to code as the following:

while (storage->removeProp("planes[@fromGroup='1']"))
{
}

because it makes it clear that the body is intentionally blank.  No need to change for this PR since it matches other examples.

Although having said that there should be a function that avoids the need for the loop.  @jakesmith Added HPCC-33048.

File: dali/base/dafdesc.cpp [hunk 7]
```diff
@@ -3498,6 +3498,11 @@ bool GroupInformation::checkIsSubset(const GroupInformation & other)
 
 void GroupInformation::createStoragePlane(IPropertyTree * storage, unsigned copy) const
 {
+    // Check that storage plane does not already have a definition
+    VStringBuffer xpath("planes[@name='%s']", name.str());
+    if (storage->hasProp(xpath))
```
Comment (ghalliday) [inline dali/base/dafdesc.cpp:3503]: In a future PR it might be interesting to allow some attributes for the implicit storage planes to be set in the configuration - so merge the definitions.  For example to be able to set expert options on thor storage planes, even if only for testing.

**Review by ghalliday — APPROVED** [review]
@jackdelv please squash.

File: dali/base/dafdesc.cpp (hunk 7 above)
Comment (jackdelv) [reply dali/base/dafdesc.cpp:3503]: I agree and have opened this [JIRA](https://hpccsystems.atlassian.net/browse/HPCC-33057).

**Comment by jackdelv** [comment]
@ghalliday Squashed.
